controllers/subscription_controller.go

```diff
@@ -18,45 +18,378 @@ package controllers
+		Watches(&source.Kind{Type: &corev1alpha1.Repository{}},
+			handler.EnqueueRequestsFromMapFunc(func(o client.Object) (reqs []reconcile.Request) {
+				return r.GetReqs(o, false)
+			})).
 		Complete(r)
 }
+
+// GetReqs get subscription reqs
+func (r *SubscriptionReconciler) GetReqs(o client.Object, watchComponent bool) (reqs []reconcile.Request) {
+	key := o.GetName() + "/" + o.GetNamespace()
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+	var subs sets.String
+	var ok bool
+	if watchComponent {
+		subs, ok = r.ComponentWatchMap[key]
+	} else {
+		subs, ok = r.RepositoryWatchMap[key]
+	}
+	if !ok {
+		return
+	}
+	all := subs.List()
+	for _, val := range all {
+		name, ns, ok := strings.Cut(val, "/")
+		if !ok {
+			continue
+		}
+		reqs = append(reqs, reconcile.Request{NamespacedName: client.ObjectKey{Name: name, Namespace: ns}})
+	}
+	return reqs
+}
+
+// SubscriptionUpdate check if subscription spec or status is updated
+func (r *SubscriptionReconciler) SubscriptionUpdate(event event.UpdateEvent) bool {
+	oldSub, ok := event.ObjectOld.(*corev1alpha1.Subscription)
+	if !ok {
+		return false
+	}
+	newSub, ok := event.ObjectNew.(*corev1alpha1.Subscription)
+	if !ok {
+		return false
+	}
+	if !reflect.DeepEqual(oldSub.Spec, newSub.Spec) {
+		return true
+	}
+	if !reflect.DeepEqual(oldSub.Status, newSub.Status) {
+		return true
+	}
+	return false
+}
+
+// SubscriptionDelete clean Reconciler cache map when subscription is deleted
+func (r *SubscriptionReconciler) SubscriptionDelete(event event.DeleteEvent) bool {
+	sub, ok := event.Object.(*corev1alpha1.Subscription)
+	if !ok {
+		return false
+	}
+	val := sub.Name + "/" + sub.Namespace
+	component := sub.Spec.ComponentRef.Name + "/" + sub.Spec.ComponentRef.Namespace
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	r.ComponentWatchMap[component].Delete(val)
+	if sub.Spec.RepositoryRef != nil {
+		repository := sub.Spec.RepositoryRef.Name + "/" + sub.Spec.RepositoryRef.Namespace
+		r.RepositoryWatchMap[repository].Delete(val)
+	}
+	return false
+}
+
+// UpdateStatusRepositoryHealth get repository CR, check if the repository is healthy and updates subscription status.RepositoryHealth
+func (r *SubscriptionReconciler) UpdateStatusRepositoryHealth(ctx context.Context, logger logr.Logger, sub *corev1alpha1.Subscription) (err error) {
+	repo := &corev1alpha1.Repository{}
+	if err = r.Get(ctx, types.NamespacedName{Namespace: sub.Spec.RepositoryRef.Namespace, Name: sub.Spec.RepositoryRef.Name}, repo); err != nil {
+		logger.Error(err, "Failed to get repository", "Repository.Namespace", sub.Spec.RepositoryRef.Namespace, "Repository.Name", sub.Spec.RepositoryRef.Name)
+		if err1 := r.PatchCondition(ctx, sub, corev1alpha1.SubscriptionReconcileError(corev1alpha1.SubscriptionTypePlanSynce, err)); err1 != nil {
+			return errors.Wrap(err, err1.Error())
+		}
+		return err
+	}
+	// FIXME check if the repository is healthy
+	healthy := true
+	if repo.Status.GetCondition(corev1alpha1.TypeReady).Status != corev1.ConditionTrue {
```

Comment:
@0xff-dev do we have health check in repository? if yes, double check on this part.

Reply:
No.

Reply:
🆗